src/plugins/upf/dpi.c

```diff
@@ -0,0 +1,1133 @@
+
+  args->id = id;
+  args->res = 1;
+
+  return 0;
+}
+
+int
+upf_dpi_lookup(u32 db_index, u8 * str, uint16_t length, u32 * app_index)
+{
+  upf_dpi_entry_t *entry = NULL;
+  int ret = 0;
+  upf_dpi_cb_args_t args = {};
+
+  if (db_index == ~0)
+    return -1;
+
+  entry = pool_elt_at_index (upf_dpi_db, db_index);
+  if (!entry)
+    return -1;
+
+  ret = hs_scan(entry->database, (const char*)str, length, 0, entry->scratch,
+                upf_dpi_event_handler, (void*)&args);
+  if (ret != HS_SUCCESS)
+    return -1;
+
+  if (args.res == 0)
+    return -1;
+
+  *app_index = args.id;
+
+  return 0;
+}
+
+int
+upf_dpi_remove(u32 db_index)
+{
+  upf_dpi_entry_t *entry = NULL;
+
+  entry = pool_elt_at_index (upf_dpi_db, db_index);
+  if (!entry)
+    return -1;
+
+  upf_dpi_cleanup_db_entry(entry);
+
+  pool_put (upf_dpi_db, entry);
+
+  return 0;
+}
+
+static void
+upf_add_rules(u32 app_index, upf_dpi_app_t *app, upf_dpi_args_t ** args, u8 path)
+{
+  u32 index = 0;
+  u32 rule_index = 0;
+  upf_dpi_rule_t *rule = NULL;
+  upf_dpi_args_t arg;
+
+  /* *INDENT-OFF* */
+  hash_foreach(rule_index, index, app->rules_by_id,
+  ({
+     rule = pool_elt_at_index(app->rules, index);
+
+     if (rule->path)
+       {
+         arg.index = app_index;
+         if (path)
+           {
+             arg.rule = rule->path;
+           }
+         else
+           {
+             arg.rule = rule->host;
+           }
+         vec_add1(*args, arg);
+       }
+  }));
+  /* *INDENT-ON* */
+}
+
+static inline void
+upf_add_path_rules(u32 app_index, upf_dpi_app_t *app, upf_dpi_args_t ** args)
+{
+  return upf_add_rules(app_index, app, args, 1);
+}
+
+static inline void
+upf_add_host_rules(u32 app_index, upf_dpi_app_t *app, upf_dpi_args_t ** args)
+{
+  return upf_add_rules(app_index, app, args, 0);
+}
+
+int
+upf_dpi_get_db_id(u8 * app_name, u32 * path_db_index, u32 * host_db_index)
+{
+  uword *p = NULL;
+  upf_main_t * sm = &upf_main;
+  upf_dpi_app_t *app = NULL;
+
+  p = hash_get_mem (sm->upf_app_by_name, app_name);
+
+  if (!p)
+    return -1;
+
+  app = pool_elt_at_index(sm->upf_apps, p[0]);
+
+  *path_db_index = app->path_db_index; 
+  *host_db_index = app->host_db_index; 
+
+  return 0;
+}
+
+static int
+upf_dpi_create_update_db(u8 * app_name, u32 * path_db_index, 
+                         u32 * host_db_index)
+{
+  uword *p = NULL;
+  upf_dpi_args_t *args = NULL;
+  upf_main_t * sm = &upf_main;
+  upf_dpi_app_t *app = NULL;
+  int res = 0;
+
+  p = hash_get_mem (sm->upf_app_by_name, app_name);
+
+  if (!p)
+    return -1;
+
+  app = pool_elt_at_index(sm->upf_apps, p[0]);
+
+  upf_add_path_rules(p[0], app, &args);
+
+  if (!args)
+    return -1;
+
+  res = upf_dpi_add_multi_regex(args, path_db_index);
+
+  vec_free(args);
+
+  upf_add_host_rules(p[0], app, &args);
+
+  if (!args)
+    return -1;
+
+  res = upf_dpi_add_multi_regex(args, host_db_index);
+
+  vec_free(args);
+
+  return res;
+}
+
+static void
+upf_dpi_all_pdr_update(u8* app_name)
```

Comment:
why is that needed?

Does dpi_{path,host}_db_id change when rules are added? If so, why?